src/sage/schemes/curves/plane_curve_arrangement.py

```diff
@@ -0,0 +1,1371 @@
+        if simplified:
+            return self._meridians_simpl
+        else:
+            return self._meridians_nonsimpl
+
+
+class PlaneCurveArrangements(UniqueRepresentation, Parent):
+    """
+    Plane curve arrangements.
+
+    INPUT:
+
+    - ``base_ring`` -- ring; the base ring
+
+    - ``names`` -- tuple of strings; the variable names
+
+    EXAMPLES::
+
+        sage: H.<x, y> = AffinePlaneCurveArrangements(QQ)
+        sage: H(x, y^2, x-1, y-1)
+        Arrangement (x, y^2, x - 1, y - 1) in Affine Space
+        of dimension 2 over Rational Field
+    """
+    Element = PlaneCurveArrangementElement
+
+    @staticmethod
+    def __classcall__(cls, base, names=()):
+        """
+        Normalize names
+        TESTS::
+
+            sage: H.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: K = AffinePlaneCurveArrangements(QQ, names=('x', 'y'))
+            sage: H is K
+            True
+            sage: TestSuite(K).run()
+        """
+        names = normalize_names(len(names), names)
+        return super().__classcall__(cls, base, names)
+
+    def __init__(self, base_ring, names=tuple()):
+        """
+        Initialize ``self``.
+
+        TESTS::
+
+            sage: H.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: TestSuite(H).run()
+            sage: H.<x, y, z> = ProjectivePlaneCurveArrangements(QQ)
+            sage: K = ProjectivePlaneCurveArrangements(QQ, names=('x', 'y', 'z'))
+            sage: H is K
+            True
+            sage: type(K)
+            <class 'sage.schemes.curves.plane_curve_arrangement.ProjectivePlaneCurveArrangements_with_category'>
+        """
+        if base_ring not in _Fields:
+            raise ValueError('base ring must be a field')
+        super().__init__(base_ring, names=names, category=Sets())
+        self._embedded = None
+        if len(names) == 2:
+            self._embedded = 'affine'
+        elif len(names) == 3:
+            self._embedded = 'projective'
+
+    def coordinate_ring(self):
+        """
+        Return the coordinate ring.
+
+        OUTPUT:
+
+        The coordinate ring of the curve arrangement.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: L.coordinate_ring()
+            Multivariate Polynomial Ring in x, y over Rational Field
+        """
+        return PolynomialRing(self.base_ring(), self.variable_names())
+
+    @abstract_method
+    def change_ring(self, base_ring):
+        """
+        Return curve arrangements over a different base ring.
+
+        INPUT:
+
+        - ``base_ring`` -- a ring; the new base ring.
+
+        OUTPUT:
+
+        A new :class:`PlaneCurveArrangements` instance over the new
+        base ring.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = PlaneCurveArrangements(QQ)
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: <abstract method ambient_space at  0x...>
```

Comment:
This test is strange. You shouldn't need this method for constructing the abstract object. Nor is this the place it fails. I would use a concrete class.

Although couldn't this be implemented generically as
```python
return type(self)(base_ring, names=self.variable_names())
```
?

Reply:
This was one of the points of the last changes. Do you propose to do not use `@abstract_method` and use the suggestion?

Reply:
No, I propose exactly what I proposed before. The `@abstract_method` only refers to the `ambient_space()` because that can*not* be implemented generically. However, in this case, you can simply use the`type()` with the new base ring. This change removes redundancy and supports the abstraction.

Reply:
There is a failed test with the change, this is a more detailed output:
```
sage: L.<x, y> = AffinePlaneCurveArrangements(QQ)
sage: L1 = L.change_ring(QQ)
sage: L.__reduce__()
(<function unreduce at 0x7ff2c7782f20>,
 (<class 'sage.schemes.curves.plane_curve_arrangement.AffinePlaneCurveArrangements'>,
  (Rational Field, ('x', 'y')),
  {}))
sage: L1.__reduce__()
(<function unreduce at 0x7ff2c7782f20>,
 (<class 'sage.schemes.curves.plane_curve_arrangement.AffinePlaneCurveArrangements_with_category'>,
  (Rational Field, ('x', 'y')),
  {}))
```